### ISSUE TITLE
Bos.OS.File.read on stdin

### ISSUE THREAD

**Issue by rizo** [dbuenzli/bos#78]
I was surprised that `Bos.OS.File.read (Fpath.v "/dev/stdin")` is an empty string while `Bos.OS.File.read (Fpath.v "-")` actually waits for input. Is that intentional?

**Comment by dbuenzli** [comment]
I don't know. There's no special casing for `/dev/stdin` it treats it like a file, I don't know how this file behaves on your system. 

`-` is special cased and reads directly from the OCaml `stdin` `in_channel` or `Unix.stdin` because that's the widespread unix convention to denote `stdout` and `stdin` where files are expected in Unix tools. See [here](http://erratique.ch/software/bos/doc/Bos.OS.File.html#VALdash) for details.

**Comment by dbuenzli** [comment]
(Note that using `/dev/stdout` for writing rather than `-` will likely fail because of `bos`'s atomic writes, see #70).

**Comment by dbuenzli** [comment]
Ah I think I know why this is the case. Files are [read](https://github.com/dbuenzli/bos/blob/88d3ba395ca8401d72e8d7c2df1c5065afd475cf/src/bos_os_file.ml#L99-L109) in one `read` and for that `in_channel_length` is used. I guess this returns `0` on `/dev/stdin`.

**Comment by rizo** [comment]
I'm aware of the `-` convention and I think it's good that Bos follows it.

I imagine people would be equally likely to refer to `/dev/stdin` directly. In fact I would expect it to support other kinds of channels that don't have a fixed length (like named pipes).

I submitted a PR to add support for `S_CHR` (_e.g._ `/dev/stdin`) and `S_FIFO` (_i.e._ named pipes) devices.

See https://github.com/ocaml/ocaml/blob/trunk/otherlibs/unix/unix.mli#L415

**Comment by dbuenzli** [comment]
I wonder if the best here is not to simply read the file in the "usual" way whenever `in_channel_length` reports `0`.

**Comment by rizo** [comment]
That would work I think.

The only edge case is when a file is actually empty – `in_channel_length` reports `0` in that case too.  The later call to `input` would produce `0`, and thus an empty string, as expected. It would unnecessarily allocating extra `IO_BUFFER_SIZE `, which is not a big deal.

**Comment by rizo** [comment]
Following your idea seems like we can indeed separate the read operation for _regular_ files and the _streaming_ ones. The following checks could be performed to make the decision:

- `in_channel_length ic > 0`: Read regular non-empty file.
- `in_channel_length ic = 0 && Unix.((stat file).st_kind = S_REG)`: Produce an empty string.
- Otherwise (_i.e._ empty non-regular file): Read streaming file (like named pipe, char device, socket, etc).

**Comment by dbuenzli** [comment]
I don't mind about the case for empty files, in practice simply making the test for zero should be enough.

**Comment by rizo** [comment]
I'll change the PR then :)

**Comment by dbuenzli** [comment]
Humpf. Then I guess it's either reimplementing `in_channel_length` using  `Unix.lseek` and catching `ESPIPE` or `stat`ing. I would favor the former.

**Comment by dbuenzli** [comment]
The code [here](https://github.com/b0-system/b0/blob/bc3a4f63e3dd6a0ea19a3725016ad64f60acb62e/src-std/b0_os.ml#L216-L236) basically does the `lseek` stuff but does not handle `ESPIPE`.

**Comment by dbuenzli** [comment]
Incidentally I had to go over this. I think something along this way (but on channels should do): 
```ocaml
    let to_string fd =
      let b = Bytes.create unix_buffer_size in
      let acc = Buffer.create unix_buffer_size in
      let rec loop () = match Unix.read fd b 0 (Bytes.length b) with
      | 0 -> Buffer.contents acc
      | l -> Buffer.add_subbytes acc b 0 l; loop ()
      | exception Unix.Unix_error (Unix.EINTR, _, _) -> loop ()
      in
      loop ()

    let rec really_read fd b start len = match len <= 0 with
    | true -> ()
    | false ->
        match Unix.read fd b start len with
        | 0 -> failwith (err_doing "Reading" "Unexpected end of file")
        | r -> really_read fd b (start + r) (len - r)
        | exception Unix.Unix_error (Unix.EINTR, _, _) ->
            really_read fd b start len

    let read_file file fd =
      try
        match Unix.lseek fd 0 Unix.SEEK_END with
        | exception Unix.Unix_error (Unix.ESPIPE, _, _) -> to_string fd
        | len when len > Sys.max_string_length ->
            failwithf "File to read too large: %d bytes, max supported: %d"
              len Sys.max_string_length
        | len ->
            let b = Bytes.create len in
            ignore (Unix.lseek fd 0 Unix.SEEK_SET);
            really_read fd b 0 len;
            Bytes.unsafe_to_string b
      with
      | Failure e -> Fmt.strf "%a: %s" Fpath.pp file e
      | Unix.Unix_error (e, _, _) ->
          Fmt.strf "%a: %s" Fpath.pp file (err_doing "Reading" (uerr e))
```

**Comment by dbuenzli** [comment]
> Sorry for the delay, I'll try to find time to try this soon.

Well no need to be sorry I'm the one who's having the bug here...

**Comment by rizo** [comment]
It's about communicating the expectation that I haven't given up on this :-P

**Comment by rizo** [comment]
Did you have a change to look at https://github.com/dbuenzli/bos/pull/79? Any suggestions/concerns?

**Comment by dbuenzli** [comment]
Thanks. Fixed by @rizo in 61526bf67f218fd730db8ac19.
